Add manager wrapper for EBPF

gjulianm opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.

-->
### What does this PR do?

This PR adds a new wrapper for the manager that is intended to generalize the existing eBPF telemetry manager, adding "modifiers" which are interfaces that can modify the manager at certain points in its lifetime. For now, these points are `BeforeInit`and `AfterInit`, although more can be added easily.

In order to add a modifier, first the interface needs to be defined with the corresponding methods and then an instance of that interface must be registered using the `RegisterModifier` function (see [this example](https://github.com/DataDog/datadog-agent/blob/75f2b328470395e155badd3e21a21a2d6adb5adf/pkg/ebpf/modifiers/printk_patcher.go#L216-L234)), preferrably in an `init` method.

Users of this manager can either choose to have all default modifiers enabled, or define exactly which ones they want to enable (also available the option to disable all modifiers). This allows the eBPF team to change the behavior of the manager without modifying all callers, but allows users to retain full control if they need to.

This PR does not modify callers of the eBPF manager, that will be done in a separate PR and once the eBPF telemetry is completely decoupled.

ebpf-manager is bumped to version 0.5.0, bringing in support for multiple instruction patchers.

### Motivation

https://datadoghq.atlassian.net/browse/EBPF-339
After implementing an instruction patcher, it was clear that an alternative solution was needed that allowed modifying the managers (in this case, adding an instruction patcher) without changing the code everywhere the manager was used. This PR will also help in the parallel efforts that are being done to decouple the eBPF telemetry from the manager.

### Additional Notes

This PR also moves the `printk_patcher*` files to `pkg/ebpf/modifiers`. A separate package has been created because it's needed to avoid import loops due to imports in `printk_patcher_test.go`, it's highly likely that other modifiers will need the same imports too.

### Possible Drawbacks / Trade-offs

The logic to decide which modifiers are used is simple: either choose the defaults or be explicit in which modifiers are used. It is possible that in the future a more complex system (allowing exclusions from the defaults) is needed, but for now we've settled on this.

### Describe how to test/QA your changes

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided. Except if the `qa/skip-qa` label, with required either `qa/done` or `qa/no-code-change` labels, are applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

gjulianm committed:
Add new manager wrapper

gjulianm committed:
Use a map for the modifiers

gjulianm committed:
standardize import name

gjulianm committed:
Update ebpf-manager to 0.5.0

gjulianm committed:
Update InstructionPatcher references

brycekahle commented on pkg/ebpf/modifiers/printk_patcher.go:
```diff
@@ -207,3 +212,23 @@ func patchPrintkInstructions(p *ebpf.ProgramSpec) (int, error) {
+}
+
+func (t *printkPatcherModifier) Name() string {
+	return ddebpf.PrintkModifier
```
I think the name constants should ideally live next to the implementations.

brycekahle commented on pkg/ebpf/manager.go:
```diff
@@ -0,0 +1,133 @@
+// NewManager creates a manager wrapper.
+// If the modifiers list is empty, it will be initialized with the default modifiers.
+// Pass manager.NO_MODIFIERS to avoid initializing the default modifiers.
+func NewManager(mgr *manager.Manager, modifiers ...string) *Manager {
```
Do we even need the strings? Could we use types instead and each modifier expose a global `var PrintKPatcher Modifier`, then the call becomes `NewManager(mgr, modifiers.PrintKPatcher)`

brycekahle replied:
I think this might remove the need to register modifiers as well. The globals could start out `nil` and then triggered to be created from `ebpf.Setup()` when we know eBPF is required?

brycekahle reviewed (changes requested):
I think there might be a simpler approach. See my comments about using exported globals in each modifier. I'd be happy to show you what I mean in a branch too, just let me know.

gjulianm commented on pkg/ebpf/modifiers/printk_patcher.go:
```diff
@@ -207,3 +212,23 @@ func patchPrintkInstructions(p *ebpf.ProgramSpec) (int, error) {
+}
+
+func (t *printkPatcherModifier) BeforeInit(m *ddebpf.Manager, _ *manager.Options) error {
+	m.InstructionPatcher = PatchPrintkNewline
```
Yes, that's ready in a separate PR (https://github.com/DataDog/datadog-agent/pull/22659), I can put everything together too if that's better

brycekahle replied:
Let's pull that into this PR. This one doesn't make much sense if you can't have multiple independent modifiers which use instruction patching.

gjulianm committed:
Eliminate dependency of printk_patcher_test on runtime package, reduce import loops

gjulianm committed:
Suggestions from PR

gjulianm committed:
Remove unused go-generate statements

gjulianm commented:
> I think there might be a simpler approach. See my comments about using exported globals in each modifier. I'd be happy to show you what I mean in a branch too, just let me know.

(responding here to all the comments together) Done, I think I got what you meant. I changed the way the `printk_patcher_test.go` file was loading the binary, with that I could remove the import loops and put everything in the `pkg/ebpf` package. I then changed the interfaces to accept `Modifier` instances instead of strings as you said, although I didn't add global variables for each modifier: I think there's not much difference between doing `NewManager(..., PrintkModifier)`and `NewManager(..., &PrintkModifier{})`, and with the second approach the modifiers can be a little bit more flexible. I also removed the `Name()` property as it was only useful for debug/error logs at that point and it could be replaced with `reflect.TypeOf()` for identification of the modifier instance.

gjulianm committed:
Comment clarifying internal state

brycekahle commented:
> I think there's not much difference between doing `NewManager(..., PrintkModifier)`and `NewManager(..., &PrintkModifier{})`

I think the only difference would be if multiple managers want to use the same modifiers, should they each get a unique instance or should they share them?

gjulianm commented:
> > I think there's not much difference between doing `NewManager(..., PrintkModifier)`and `NewManager(..., &PrintkModifier{})`
> 
> I think the only difference would be if multiple managers want to use the same modifiers, should they each get a unique instance or should they share them?

I'm not sure but I'm leaning towards allowing a unique instance for each manager. In terms of memory it isn't too much, and it also opens the door to modifiers that have options to change behavior or even per-manager internal state.

brycekahle commented on pkg/ebpf/manager.go:
```diff
@@ -0,0 +1,83 @@
+// InitWithOptions is a wrapper around ebpf-manager.Manager.InitWithOptions
+func (m *Manager) InitWithOptions(bytecode io.ReaderAt, opts *manager.Options) error {
+	for _, mod := range m.EnabledModifiers {
+		modName := reflect.TypeOf(mod).String()
```
I think it would probably be better to implement `String()` on the modifier types which can just return a constant string, rather than using reflection which is slow. If you embed `fmt.Stringer` in the `Modifier` interface, then you can make sure folks don't forget. 😄

brycekahle commented on pkg/ebpf/manager.go:
```diff
@@ -0,0 +1,83 @@
+func (m *Manager) InitWithOptions(bytecode io.ReaderAt, opts *manager.Options) error {
+	for _, mod := range m.EnabledModifiers {
+		modName := reflect.TypeOf(mod).String()
+		log.Debugf("Running %s manager modifier", modName)
```
once you have `.String()`, you can do:
```suggestion
		log.Debugf("Running %s manager modifier", mod)
```

brycekahle commented on pkg/ebpf/printk_patcher.go:
```diff
@@ -207,3 +207,18 @@ func patchPrintkInstructions(p *ebpf.ProgramSpec) (int, error) {
+}
+
+// BeforeInit adds the PatchPrintkNewline function to the manager
+func (t *PrintkPatcherModifier) BeforeInit(m *Manager, _ *manager.Options) error {
```
I still think it is probably a good idea to use `*manager.Manager` here and in `AfterInit`, so modifiers don't have the ability to mess with the wrapper manager.

brycekahle commented on pkg/ebpf/ebpf.go:
```diff
@@ -56,11 +58,19 @@ func coreLoader(cfg *Config) (*coreAssetLoader, error) {
-// Reset resets CO-RE and BTF loaders back to uninitialized state
+func registerDefaultModifiers() {
+	defaultModifiers = []Modifier{
+		&PrintkPatcherModifier{},
```
Noting that managers that use the default modifiers will all share an instance.

gjulianm committed:
Add fmt.Stringer and change to use manager.Manager in BeforeInit/AfterInit

gjulianm committed:
Comments regarding sharing instances

brycekahle commented on pkg/ebpf/printk_patcher.go:
```diff
@@ -207,3 +207,22 @@ func patchPrintkInstructions(p *ebpf.ProgramSpec) (int, error) {
+
+// BeforeInit adds the PatchPrintkNewline function to the manager
+func (t *PrintkPatcherModifier) BeforeInit(m *manager.Manager, _ *manager.Options) error {
+	m.InstructionPatchers = append(m.InstructionPatchers, PatchPrintkNewline)
```
nit: does `PatchPrintkNewline` need to be exported anymore?

gjulianm replied:
Yes, in case someone wants to use it directly instead of relying on the default modifiers.

brycekahle replied:
I'm not talking about the modifier, but the patching function itself

gjulianm replied:
Oh my mistake, sorry. That one doesn't need to be exported, I'll change it now.

gjulianm committed:
Do not export PatchPrintkNewline

gjulianm committed:
Fix import

gjulianm committed:
Merge branch 'main' into guillermo.julian/manager-modifier

gjulianm commented:
/merge
